adding_to_the_RDynLib_library.qmd

```diff
@@ -55,47 +54,33 @@ table(msLevel(s))
 
 this function takes as input the XcmsExperiment object, then it extract the MS2
 spectra using "featureSpectra()" function, after that we extract from the
-initial object only the MS3 fragments that have a parent ion in the result of 
+initial object only the MS3 fragments that have a parent ion in the result of
 "featureSpectra()" function, and then we do the same process for MS4.
 
 ```{r}
 ftms_all_levels <- function(ftms) {
-  
+
   ms2 <- featureSpectra(ftms, msLevel = 2)
   cat("Number of MS2 spectra found:", length(ms2), "\n")
-  
-  combined_spectra_list <- list() 
-  
-  for (origin in unique(spectraData(ms2)$dataOrigin)) {
+
+  res <- lapply(unique(ms2$dataOrigin), function(origin) {
     ms2_subset <- ms2[ms2$dataOrigin == origin]
-    
-    ms3 <- filterMsLevel(spectra(ftms), msLevel = 3)
-    ms3_filtered <- ms3[
-      ms3$dataOrigin == origin &
-      ms3$precScanNum %in% ms2_subset$scanIndex
-    ]
-    
-    ms4 <- filterMsLevel(spectra(ftms), msLevel = 4)
-    ms4_filtered <- ms4[
-      ms4$dataOrigin == origin &
-      ms4$precScanNum %in% ms3_filtered$scanIndex
-    ]
-    
-    # Combine spectra for this origin
-    combined_origin_spectra <- c(ms2_subset, ms3_filtered, ms4_filtered)
-    
-    # Only add if not empty
-    if (length(combined_origin_spectra) > 0) {
-      combined_spectra_list[[length(combined_spectra_list) + 1]] <-
-        combined_origin_spectra
-    }
-  }
-  
 
-    final_feature_data <- do.call(c, combined_spectra_list)
+    ms3_filtered <- filterDataOrigin(filterMsLevel(spectra(ftms), 3), origin)
+    ## To support n:m matches
+    m <- findMatches(ms2_subset$acquisitionNum, ms3_filtered$precScanNum)
+    ms3_filtered <- ms3_filtered[to(m)]
+    ms3_filtered$feature_id <- ms2_subset$feature_id[from(m)]
```

Comment:
the other advantage of the `findMatches()` is that we can extract the `feature_id` of the MS2 and assign it (at the correct places) to the MS3 spectra.